### ISSUE TITLE
hcastle: wrong RAM mapping (Haunted Castle)

### ISSUE THREAD

**Issue by jotego** [mamedev/mame#11193]
The current way of mapping the RAM is incorrect:

https://github.com/mamedev/mame/blob/4d9ceafa16a8d77e4333498eb7c91234b9950906/src/mame/konami/hcastle.cpp#L323

according to the schematics and taking into account the [PAL dump](https://wiki.pldarchive.co.uk/index.php?title=Haunted_Castle), the RAM goes from 0x800 to 0x1FFF. But from 0x800 to 0xFFF the address bit 11 is set by the **WORKBK** signal. The **WORKBK** signal is the data bit 5 written here:

https://github.com/mamedev/mame/blob/4d9ceafa16a8d77e4333498eb7c91234b9950906/src/mame/konami/hcastle.cpp#L287

This is a snapshot from the schematics where the **WORKBK** and **WORKCS** signals are visible:

![imagen](https://user-images.githubusercontent.com/1863036/236511134-bb25191a-b6dc-44f1-bff3-78e80ef68ca9.png)

The PAL equation for **/WORKCS** is 

```
/WORKCS = /NAS & /A15 & /A14 & /A13 & /A12 & A11 +
       /NAS & /A15 & /A14 & /A13 & A12
```

Where `/NAS` is an address strobe. 

I have built this on FPGA and the game works normally. The software _may not_ be using this hardware feature. I still suggest implementing it in the driver for documentation purposes.

**Comment by rb6502** [comment]
Ahh, interesting!  That's where the other 2K of that chip went, and done in the most Konami way possible.  Are the OUT1/OUT2 signals in that same location connected anywhere?

**Comment by jotego** [comment]
Yes, those drive the coin counters

**Comment by rb6502** [comment]
One more question: MAME currently has RAM from 0x700-0x7FF, which isn't used normally but it is read and written a lot at the end of stage 1 (when the bricks start flying at you).  The game seems to work fine with nothing there, so it may well be a bug in the game's program.  Is anything supposed to be there?

**Comment by jotego** [comment]
As far as I can tell, that address range (700-7FF) does not select any device.

**Comment by rb6502** [comment]
Thanks again!  After further testing only version E accesses that region, so it's a software bug that got fixed.

**Comment by rb6502** [comment]
Fixed in eb2ebc1.
